src/OpenLoco/src/GameCommands/CreateTown.cpp

```diff
@@ -0,0 +1,200 @@
+using namespace OpenLoco::World;
+
+namespace OpenLoco::GameCommands
+{
+    // 0x00496FE7
+    // TODO: only used here, but move to town manager?
+    static Town* initialiseTown(Pos2 pos)
+    {
+        registers regs;
+        regs.eax = pos.x;
+        regs.ecx = pos.y;
+        call(0x00496FE7, regs);
+
+        if (regs.esi != -1)
+            return reinterpret_cast<Town*>(regs.esi);
+        else
+            return nullptr;
+    }
+
+    // 0x004C5604
+    // TODO: used by other functions; move elsewhere
+    static uint16_t sub_4C5604(Pos2 pos)
+    {
+        registers regs;
+        regs.ax = pos.x;
+        regs.cx = pos.y;
+        call(0x004C5604, regs);
+        return regs.dx;
+    }
+
+    static bool checkSurroundings(Pos2 pos)
+    {
+        auto tile = TileManager::get(pos);
+        auto* surfaceEl = tile.surface();
+        if (surfaceEl->slope() != 0 || surfaceEl->water() != 0)
+        {
+            return false;
+        }
+
+        auto* landObj = ObjectManager::get<LandObject>(surfaceEl->terrain());
+        if (landObj->hasFlags(LandObjectFlags::isDesert | LandObjectFlags::noTrees))
+        {
+            auto dx = sub_4C5604(pos);
+            if (dx < 10 && getGameState().rng.randNext() & 0xFF)
+            {
+                return false;
+            }
+        }
+
+        for (auto& town : TownManager::towns())
+        {
+            auto manhattanDistance = Math::Vector::manhattanDistance(pos, Pos2(town.x, town.y));
+            if (manhattanDistance < 768)
+            {
+                setErrorText(StringIds::too_close_to_another_town);
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    // 0x00496C22
+    static uint32_t createTown(const TownPlacementArgs& args, const uint8_t flags)
+    {
+        auto& gameState = getGameState();
+        Pos2 pos = args.pos;
+        Town* newTown = nullptr;
+
+        if (args.pos.x == -1)
+        {
+            bool foundPos = false;
+            for (auto attempts = 200; attempts > 0; attempts--)
+            {
+                uint32_t rand = gameState.rng.randNext();
+                auto tilePos = TilePos2(((rand >> 16) * kMapColumns) >> 16, ((rand & 0xFFFF) * kMapRows) >> 16);
+                pos = toWorldSpace(tilePos);
+
+                if (pos.x < 384 || pos.y < 384 || pos.x > 11904 || pos.y > 11904)
+                {
+                    continue;
+                }
+
+                if (checkSurroundings(pos))
+                {
+                    foundPos = true;
+                    break;
+                }
+            }
+
+            if (!foundPos)
+            {
+                return FAILURE;
+            }
+        }
+        else
+        {
+            bool foundPos = false;
+            for (int attempts = 0; attempts < 40; attempts++)
+            {
+                if (attempts == 0)
+                {
+                    uint32_t rand = gameState.rng.randNext();
```

Comment:
this doesn't look correct. attempt 0 is a just use the current x,y. Attempts 1 -> 10 is the & 0x60. And attempt 11 > 39 is the & 0xE0.